File: ts/src/filesystem.ts
import { gzipSync } from 'node:zlib'

import { DataPlaneClient, withQuery } from './transport.js'
import { FileNotFoundError, InvalidArgumentError } from './errors.js'
import { ProcessFrame, ProcessSocket, base64Encode } from './processSocket.js'

export enum FileType {
  /** Regular file. */
  FILE = 'file',
  /** Directory. */
  DIR = 'dir',
  /** Symbolic link. */
  SYMLINK = 'symlink',
}

export enum FilesystemEventType {
  CHMOD = 'chmod',
  CREATE = 'create',
  REMOVE = 'remove',
  RENAME = 'rename',
  WRITE = 'write',
}

/** Metadata for one sandbox filesystem entry. */
export interface EntryInfo {
  name: string
  type: FileType | string
  path: string
  size?: number
  mode?: number
  permissions?: string
  owner?: string
  group?: string
  modifiedTime?: Date
  symlinkTarget?: string
  uid?: number
  gid?: number
  mtime?: number
  metadata?: Record<string, string>
}

export type WriteInfo = EntryInfo

export type WriteData = string | Uint8Array | ArrayBuffer | Blob | ReadableStream<Uint8Array>

export interface WriteEntry {
  path: string
  data: WriteData
}

export interface FilesystemEvent {
  name: string
  type: FilesystemEventType | string
  path: string
  entry?: EntryInfo
  raw: Record<string, unknown>
}

export interface WatchOpts {
  recursive?: boolean
  includeEntry?: boolean
  allowNetworkMounts?: boolean
  timeoutMs?: number
  requestTimeoutMs?: number
  signal?: AbortSignal
  user?: string
  onExit?: (error?: Error) => void | Promise<void>
}

export interface FilesystemRequestOpts {
  requestTimeoutMs?: number
  signal?: AbortSignal
  user?: string
}

export interface FilesystemReadOpts extends FilesystemRequestOpts {
  gzip?: boolean
}

export interface FilesystemWriteOpts extends FilesystemRequestOpts {
  gzip?: boolean
  useOctetStream?: boolean
  metadata?: Record<string, string>
}

/** Live filesystem watcher. Call `stop()` to close the local watch stream. */
export class WatchHandle {
  private readonly done: Promise<void>

  constructor(
    private readonly socket: ProcessSocket,
    events: AsyncIterable<ProcessFrame>,
    onEvent: (event: FilesystemEvent) => void | Promise<void>,
    onExit?: (error?: Error) => void | Promise<void>
  ) {
    this.done = this.pump(events, onEvent, onExit)
  }

  /** Stop watching the directory. */
  stop(): void {
    this.socket.close()
  }

  /** Resolves when the watcher stream exits. */
  wait(): Promise<void> {
    return this.done
  }

  private async pump(
    events: AsyncIterable<ProcessFrame>,
    onEvent: (event: FilesystemEvent) => void | Promise<void>,
    onExit?: (error?: Error) => void | Promise<void>
  ): Promise<void> {
    let error: Error | undefined
    try {
      for await (const frame of events) {
        if (frame.type !== 'events' || !Array.isArray(frame.events)) continue
        for (const item of frame.events) {
          await onEvent(filesystemEvent(item))
        }
      }
    } catch (caught) {
      error = caught instanceof Error ? caught : new Error(String(caught))
      throw error
    } finally {
      await onExit?.(error)
    }
  }
}

/** Lazy filesystem watcher. Add listeners, then call `start()`. */
export class FilesystemWatcher {
  private handle?: WatchHandle
  private listeners: Array<(event: FilesystemEvent) => void | Promise<void>> = []

  constructor(
    private readonly dataPlane: DataPlaneClient,
    private readonly path: string,
    private readonly opts: WatchOpts = {}
  ) {}

  async start(opts: WatchOpts = {}): Promise<void> {
    if (this.handle) return
    const nextOpts = { ...this.opts, ...opts }
    const socket = await new ProcessSocket(
      this.dataPlane.baseUrl,
      this.dataPlane.token,
      withQuery('/runtime/v1/files/watch', {
        path: this.path,
        recursive: nextOpts.recursive ?? false,
        include_entry: nextOpts.includeEntry,
        allow_network_mounts: nextOpts.allowNetworkMounts,
      }),
      nextOpts.requestTimeoutMs ?? nextOpts.timeoutMs,
      this.dataPlane.headers
    ).connect()
    this.handle = new WatchHandle(socket, socket, async (event) => {
      for (const listener of this.listeners) await listener(event)
    }, nextOpts.onExit)
  }

  async stop(): Promise<void> {
    this.handle?.stop()
  }

  addEventListener(listener: (event: FilesystemEvent) => void | Promise<void>): () => boolean {
    this.listeners.push(listener)
    return () => {
      const index = this.listeners.indexOf(listener)
      if (index === -1) return false
      this.listeners.splice(index, 1)
      return true
    }
  }

  wait(): Promise<void> {
    return this.handle?.wait() ?? Promise.resolve()
  }
}

/** Filesystem helper for a sandbox data-plane session. */
export class Filesystem {
  constructor(private readonly dataPlane: DataPlaneClient) {}

  /** Read file content as text, bytes, a `Blob`, or a `ReadableStream`. */
  async read(
    path: string,
    opts?: FilesystemReadOpts & { format?: 'text' }
  ): Promise<string>
  async read(
    path: string,
    opts: FilesystemReadOpts & { format: 'bytes' }
  ): Promise<Uint8Array>
  async read(
    path: string,
    opts: FilesystemReadOpts & { format: 'blob' }
  ): Promise<Blob>
  async read(
    path: string,
    opts: FilesystemReadOpts & { format: 'stream' }
  ): Promise<ReadableStream<Uint8Array>>
  async read(
    path: string,
    opts: FilesystemReadOpts & { format?: 'text' | 'bytes' | 'blob' | 'stream' } = {}
  ): Promise<string | Uint8Array | Blob | ReadableStream<Uint8Array>> {
    const bytes = await this.dataPlane.getBytes(withQuery('/runtime/v1/files', { path, gzip: opts.gzip }), opts)
    if (opts.format === 'bytes') return bytes
    if (opts.format === 'blob') return new Blob([toArrayBuffer(bytes)])
    if (opts.format === 'stream') return new ReadableStream({ start(controller) { controller.enqueue(bytes); controller.close() } })
    return new TextDecoder().decode(bytes)
  }

  /** Read a file as raw bytes. */
  async readBytes(path: string, opts: FilesystemReadOpts = {}): Promise<Uint8Array> {
    return this.read(path, { ...opts, format: 'bytes' }) as Promise<Uint8Array>
  }

  /** Write UTF-8 text, bytes, browser data objects, or a batch of file entries. */
  async write(path: string, data: WriteData, opts?: FilesystemWriteOpts): Promise<WriteInfo>
  async write(files: WriteEntry[], opts?: FilesystemWriteOpts): Promise<WriteInfo[]>
  async write(
    pathOrFiles: string | WriteEntry[],
    dataOrOpts?: WriteData | FilesystemWriteOpts,
    opts: FilesystemWriteOpts = {}
  ): Promise<WriteInfo | WriteInfo[]> {
    if (Array.isArray(pathOrFiles)) {
      return this.writeFiles(pathOrFiles, dataOrOpts as FilesystemWriteOpts | undefined)
    }

    const body = maybeGzip(await writeDataToBytes(dataOrOpts as WriteData), opts.gzip)
    const payload = await this.dataPlane.putJson(
      withQuery('/runtime/v1/files', { path: pathOrFiles, gzip: opts.gzip }),
      body,
      requestOpts(opts, opts.gzip ? { 'content-encoding': 'gzip' } : undefined)
    )
    return entryInfo(payload.file ?? payload)
  }

  /** Write raw bytes to a file. */
  async writeBytes(path: string, data: Uint8Array | ArrayBuffer, opts: FilesystemWriteOpts = {}): Promise<WriteInfo> {
    return this.write(path, data, opts)
  }

  /** Write several files in one runtime API call. */
  async writeFiles(files: WriteEntry[], opts: FilesystemWriteOpts = {}): Promise<WriteInfo[]> {
    if (files.length === 0) return []
    const payload = await this.dataPlane.postJson('/runtime/v1/files/write_files', {
      ...requestOpts(opts),
      json: {
        files: await Promise.all(files.map(async (file) => ({
          path: file.path,
          data_base64: base64Encode(maybeGzip(await writeDataToBytes(file.data), opts.gzip)),
          ...(opts.gzip ? { gzip: true } : {}),
        }))),
      },
    })
    const written = Array.isArray(payload.files) ? payload.files : []
    return written.map(entryInfo)
  }

  /** List directory entries below `path`. */
  async list(path: string, opts: FilesystemRequestOpts & { depth?: number } = {}): Promise<EntryInfo[]> {
    const payload = await this.dataPlane.getJson(withQuery('/runtime/v1/directories', { path, depth: opts.depth }), opts)
    const entries = Array.isArray(payload.entries) ? payload.entries : []
    return entries.map(entryInfo)
  }

  /** Return whether a file or directory exists at `path`. */
  async exists(path: string, opts: FilesystemRequestOpts = {}): Promise<boolean> {
    try {
      await this.getInfo(path, opts)
      return true
    } catch (error) {
      if (error instanceof FileNotFoundError) return false
      throw error
    }
  }

  /** Return stat metadata for `path`. */
  async getInfo(path: string, opts: FilesystemRequestOpts = {}): Promise<EntryInfo> {
    const payload = await this.dataPlane.getJson(withQuery('/runtime/v1/files/stat', { path }), opts)
    return entryInfo(payload.file ?? payload.entry ?? payload)
  }

  /** Remove a file at `path`. */
  async remove(path: string, opts: FilesystemRequestOpts = {}): Promise<void> {
    await this.dataPlane.deleteJson(withQuery('/runtime/v1/files', { path }), opts)
  }

  /** Move or rename a file. */
  async rename(oldPath: string, newPath: string, opts: FilesystemRequestOpts = {}): Promise<EntryInfo> {
    const payload = await this.dataPlane.postJson('/runtime/v1/files/move', {
      ...opts,
      json: { from_path: oldPath, to_path: newPath },
    })
    return entryInfo(payload.file ?? payload)
  }

  /** Create a directory. */
  async makeDir(path: string, opts: FilesystemRequestOpts = {}): Promise<boolean> {
    await this.dataPlane.postJson(withQuery('/runtime/v1/directories', { path }), opts)
    return true
  }

  /** Start watching a directory for filesystem events. */
  watchDir(path: string): FilesystemWatcher
  watchDir(
    path: string,
    onEvent: (event: FilesystemEvent) => void | Promise<void>,
    opts?: WatchOpts
  ): Promise<FilesystemWatcher>
  watchDir(
    path: string,
    onEvent?: (event: FilesystemEvent) => void | Promise<void>,
    opts: WatchOpts = {}
  ): FilesystemWatcher | Promise<FilesystemWatcher> {
    const watcher = new FilesystemWatcher(this.dataPlane, path, opts)
    if (!onEvent) return watcher
    watcher.addEventListener(onEvent)
    return watcher.start().then(() => watcher)
  }

}

async function writeDataToBytes(data: WriteData): Promise<Uint8Array> {
  if (typeof data === 'string') return new TextEncoder().encode(data)
  if (data instanceof Uint8Array) return data
  if (data instanceof ArrayBuffer) return new Uint8Array(data)
  if (typeof Blob !== 'undefined' && data instanceof Blob) return new Uint8Array(await data.arrayBuffer())
  if (isReadableStream(data)) return readStreamToBytes(data)
  throw new InvalidArgumentError(`Unsupported file data type: ${Object.prototype.toString.call(data)}`)
}

function requestOpts(opts: FilesystemRequestOpts, headers?: Record<string, string>) {
  const out: { requestTimeoutMs?: number; signal?: AbortSignal; headers?: Record<string, string> } = {}
  if (opts.requestTimeoutMs !== undefined) out.requestTimeoutMs = opts.requestTimeoutMs
  if (opts.signal !== undefined) out.signal = opts.signal
  if (headers !== undefined) out.headers = headers
  return out
}

function maybeGzip(bytes: Uint8Array, enabled?: boolean): Uint8Array {
  return enabled ? new Uint8Array(gzipSync(bytes)) : bytes
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  const buffer = new ArrayBuffer(bytes.byteLength)
  new Uint8Array(buffer).set(bytes)
  return buffer
}

function isReadableStream(value: unknown): value is ReadableStream<Uint8Array> {
  return Boolean(value && typeof value === 'object' && typeof (value as ReadableStream<Uint8Array>).getReader === 'function')
}

async function readStreamToBytes(stream: ReadableStream<Uint8Array>): Promise<Uint8Array> {
  const reader = stream.getReader()
  const chunks: Uint8Array[] = []
  let total = 0
  try {
    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      if (!(value instanceof Uint8Array)) {
        throw new InvalidArgumentError('ReadableStream file data must yield Uint8Array chunks')
      }
      chunks.push(value)
      total += value.byteLength
    }
  } finally {
    reader.releaseLock()
  }

  const bytes = new Uint8Array(total)
  let offset = 0
  for (const chunk of chunks) {
    bytes.set(chunk, offset)
    offset += chunk.byteLength
  }
  return bytes
}

function entryInfo(value: unknown): EntryInfo {
  const item = value && typeof value === 'object' ? value as Record<string, unknown> : {}
  return {
    name: String(item.name ?? ''),
    type: String(item.type ?? FileType.FILE),
    path: String(item.path ?? ''),
    size: numberValue(item.bytes ?? item.size),
    mode: numberValue(item.mode),
    permissions: stringValue(item.permissions),
    owner: stringValue(item.owner),
    group: stringValue(item.group),
    modifiedTime: dateValue(item.modified_time ?? item.modifiedTime ?? item.mtime ?? item.updated_at ?? item.updatedAt),
    symlinkTarget: stringValue(item.symlink_target ?? item.symlinkTarget),
    uid: numberValue(item.uid),
    gid: numberValue(item.gid),
    mtime: numberValue(item.mtime),
    metadata: recordOfStrings(item.metadata),
  }
}

function numberValue(value: unknown): number | undefined {
  return typeof value === 'number' ? value : undefined
}

function stringValue(value: unknown): string | undefined {
  if (typeof value === 'string') return value
  if (typeof value === 'number') return String(value)
  return undefined
}

function dateValue(value: unknown): Date | undefined {
  if (value instanceof Date) return value
  if (typeof value === 'string' || typeof value === 'number') {
    const date = new Date(value)
    if (!Number.isNaN(date.getTime())) return date
  }
  return undefined
}

function recordOfStrings(value: unknown): Record<string, string> | undefined {
  if (!value || typeof value !== 'object') return undefined
  return Object.fromEntries(Object.entries(value as Record<string, unknown>).map(([key, item]) => [key, String(item)]))
}

function filesystemEvent(value: unknown): FilesystemEvent {
  const item = value && typeof value === 'object' ? value as Record<string, unknown> : {}
  const path = String(item.path ?? '')
  return {
    type: normalizeEventType(String(item.type ?? 'modify')),
    name: String(item.name ?? relativeName(path)),
    path,
    entry: item.file && typeof item.file === 'object' ? entryInfo(item.file) : undefined,
    raw: item,
  }
}

function normalizeEventType(value: string): FilesystemEvent['type'] {
  if (value === 'delete') return FilesystemEventType.REMOVE
  if (value === 'modify') return FilesystemEventType.WRITE
  if (value === 'chmod') return FilesystemEventType.CHMOD
  if (value === 'create') return FilesystemEventType.CREATE
  if (value === 'remove') return FilesystemEventType.REMOVE
  if (value === 'rename') return FilesystemEventType.RENAME
  if (value === 'write') return FilesystemEventType.WRITE
  return value
}

function relativeName(path: string): string {
  return path.split('/').filter(Boolean).pop() ?? ''
}

File: rust/src/filesystem.rs
use serde_json::Value;

use crate::error::{Error, Result};
use crate::process_socket::ProcessSocket;
use crate::transport::DataPlaneClient;
use base64::Engine;

/// File type returned by sandbox filesystem metadata.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum FileType {
    /// Regular file.
    File,
    /// Directory.
    Dir,
    /// Symbolic link.
    Symlink,
    /// Any file type not recognized by this SDK version.
    Other(String),
}

/// Metadata for one sandbox filesystem entry.
#[derive(Clone, Debug, Default)]
pub struct EntryInfo {
    /// Basename of the entry.
    pub name: String,
    /// Absolute sandbox path.
    pub path: String,
    /// File type, when returned by the runtime.
    pub file_type: Option<FileType>,
    /// Size in bytes, when known.
    pub size: Option<u64>,
    /// Runtime metadata map.
    pub metadata: serde_json::Map<String, Value>,
}

/// Metadata returned by write operations.
pub type WriteInfo = EntryInfo;

/// File path and bytes used by `Filesystem::write_files`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WriteEntry {
    /// Absolute sandbox path to write.
    pub path: String,
    /// File content bytes.
    pub data: Vec<u8>,
}

impl WriteEntry {
    /// Create a file write entry.
    pub fn new(path: impl Into<String>, data: impl AsRef<[u8]>) -> Self {
        Self {
            path: path.into(),
            data: data.as_ref().to_vec(),
        }
    }
}

/// Options for bounded file reads.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct FileReadOptions {
    /// Maximum bytes to read before failing. When omitted, the full file is read.
    pub max_bytes: Option<usize>,
}

/// Options for applying a patch payload inside the sandbox.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffOptions {
    /// Optional working directory used to resolve relative patch paths.
    pub cwd: Option<String>,
}

/// Failed hunk metadata returned by `Filesystem::apply_diff`.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffFailedHunk {
    /// One-based hunk index within the failed file patch.
    pub index: usize,
    /// Original starting line from the failed hunk.
    pub old_start: usize,
}

/// One failed file patch returned by `Filesystem::apply_diff`.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffFailure {
    /// Path that failed to apply.
    pub path: String,
    /// Human-readable failure reason.
    pub error: String,
    /// Failed hunk details when available.
    pub failed_hunk: Option<ApplyDiffFailedHunk>,
}

/// Per-file patch summary returned by `Filesystem::apply_diff`.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffFileSummary {
    /// Target path.
    pub path: String,
    /// Source path for move/rename patches.
    pub source_path: Option<String>,
    /// Patch kind such as `created`, `updated`, or `deleted`.
    pub kind: String,
    /// Added line count.
    pub added: usize,
    /// Removed line count.
    pub removed: usize,
}

/// Aggregate patch summary returned by `Filesystem::apply_diff`.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffSummary {
    /// Number of file patches requested.
    pub requested: usize,
    /// Number of file patches applied.
    pub applied: usize,
    /// Number of file patches that failed.
    pub failed: usize,
}

/// Structured result returned by `Filesystem::apply_diff`.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ApplyDiffReport {
    /// Overall status: `applied`, `partial`, or `failed`.
    pub status: String,
    /// Number of parsed diff blocks.
    pub parsed_diff_blocks: usize,
    /// Number of parsed file patches.
    pub patches: usize,
    /// Per-file patch summaries.
    pub files: Vec<ApplyDiffFileSummary>,
    /// Aggregate patch summary.
    pub summary: ApplyDiffSummary,
    /// Paths successfully applied.
    pub applied: Vec<String>,
    /// Per-file failures.
    pub failed: Vec<ApplyDiffFailure>,
    /// Paths touched by the operation.
    pub touched: Vec<String>,
    /// Raw runtime response.
    pub raw: Value,
}

/// Filesystem event returned by a watch stream.
#[derive(Clone, Debug, Default)]
pub struct FilesystemEvent {
    /// Event type such as `create`, `write`, `remove`, or `rename`.
    pub event_type: String,
    /// Absolute sandbox path.
    pub path: String,
    /// Basename of the changed path.
    pub name: String,
    /// Entry metadata when the runtime can still stat the changed path.
    pub entry: Option<EntryInfo>,
    /// Full raw event payload.
    pub raw: Value,
}

/// Options for `Filesystem::watch_dir`.
#[derive(Clone, Debug, Default)]
pub struct WatchOptions {
    /// Whether to watch recursively.
    pub recursive: bool,
    /// Whether to include entry metadata when available.
    pub include_entry: bool,
}

/// Live filesystem watch stream.
pub struct WatchHandle {
    socket: ProcessSocket,
}

impl WatchHandle {
    /// Read the next batch of filesystem events.
    pub async fn next_events(&mut self) -> Result<Option<Vec<FilesystemEvent>>> {
        while let Some(frame) = self.socket.next_frame().await? {
            if frame.get("type").and_then(Value::as_str) != Some("events") {
                continue;
            }
            let events = frame
                .get("events")
                .and_then(Value::as_array)
                .cloned()
                .unwrap_or_default()
                .into_iter()
                .map(filesystem_event)
                .collect();
            return Ok(Some(events));
        }
        Ok(None)
    }

    /// Stop watching the directory.
    pub async fn stop(&mut self) -> Result<()> {
        self.socket.close().await
    }
}

/// Filesystem helper for a sandbox data-plane session.
#[derive(Clone)]
pub struct Filesystem {
    data_plane: DataPlaneClient,
}

impl Filesystem {
    pub(crate) fn new(data_plane: DataPlaneClient) -> Self {
        Self { data_plane }
    }

    /// Read a file as UTF-8 text.
    pub async fn read_text(&self, path: &str) -> Result<String> {
        let bytes = self.read_bytes(path).await?;
        Ok(String::from_utf8_lossy(&bytes).into_owned())
    }

    /// Read a file as raw bytes.
    pub async fn read_bytes(&self, path: &str) -> Result<Vec<u8>> {
        self.read_bytes_with_options(path, FileReadOptions::default())
            .await
    }

    /// Read a file as raw bytes with an optional SDK-side byte cap.
    pub async fn read_bytes_with_options(
        &self,
        path: &str,
        opts: FileReadOptions,
    ) -> Result<Vec<u8>> {
        self.data_plane
            .get_bytes_with_limit(
                &format!("/runtime/v1/files?path={}", urlencoding(path)),
                opts.max_bytes,
            )
            .await
    }

    /// Write bytes or text to a file.
    pub async fn write(&self, path: &str, data: impl AsRef<[u8]>) -> Result<WriteInfo> {
        let payload = self
            .data_plane
            .put_bytes(
                &format!("/runtime/v1/files?path={}", urlencoding(path)),
                data.as_ref().to_vec(),
            )
            .await?;
        Ok(entry_info(payload.get("file").unwrap_or(&payload)))
    }

    /// Write several files in one runtime API call.
    pub async fn write_files(&self, files: Vec<WriteEntry>) -> Result<Vec<WriteInfo>> {
        if files.is_empty() {
            return Ok(vec![]);
        }
        let payload = self
            .data_plane
            .post_json("/runtime/v1/files/write_files", write_files_payload(&files))
            .await?;
        Ok(payload
            .get("files")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .map(|value| entry_info(&value))
            .collect())
    }

    /// List directory entries below `path`.
    pub async fn list(&self, path: &str) -> Result<Vec<EntryInfo>> {
        let payload = self
            .data_plane
            .get_json(&format!(
                "/runtime/v1/directories?path={}",
                urlencoding(path)
            ))
            .await?;
        Ok(payload
            .get("entries")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .map(|value| entry_info(&value))
            .collect())
    }

    /// Return whether a file or directory exists at `path`.
    pub async fn exists(&self, path: &str) -> Result<bool> {
        match self.get_info(path).await {
            Ok(_) => Ok(true),
            Err(Error::FileNotFound(_)) | Err(Error::NotFound(_)) => Ok(false),
            Err(error) => Err(error),
        }
    }

    /// Return stat metadata for `path`.
    pub async fn get_info(&self, path: &str) -> Result<EntryInfo> {
        let payload = self
            .data_plane
            .get_json(&format!(
                "/runtime/v1/files/stat?path={}",
                urlencoding(path)
            ))
            .await?;
        Ok(entry_info(
            payload
                .get("file")
                .or_else(|| payload.get("entry"))
                .unwrap_or(&payload),
        ))
    }

    /// Remove a file at `path`.
    pub async fn remove(&self, path: &str) -> Result<()> {
        self.data_plane
            .delete_json(&format!("/runtime/v1/files?path={}", urlencoding(path)))
            .await?;
        Ok(())
    }

    /// Move or rename a file.
    pub async fn rename(&self, old_path: &str, new_path: &str) -> Result<EntryInfo> {
        let payload = self
            .data_plane
            .post_json(
                "/runtime/v1/files/move",
                serde_json::json!({"from_path": old_path, "to_path": new_path}),
            )
            .await?;
        Ok(entry_info(payload.get("file").unwrap_or(&payload)))
    }

    /// Apply a git-style unified diff or Codex apply_patch payload inside the sandbox.
    pub async fn apply_diff(&self, diff: &str, opts: ApplyDiffOptions) -> Result<ApplyDiffReport> {
        let payload = self
            .data_plane
            .post_json(
                "/runtime/v1/files/apply_diff",
                apply_diff_payload(diff, &opts),
            )
            .await?;
        Ok(apply_diff_report(payload))
    }

    /// Create a directory.
    pub async fn make_dir(&self, path: &str) -> Result<bool> {
        self.data_plane
            .post_json(
                &format!("/runtime/v1/directories?path={}", urlencoding(path)),
                serde_json::json!({}),
            )
            .await?;
        Ok(true)
    }

    /// Start watching a directory for filesystem events.
    pub async fn watch_dir(&self, path: &str, opts: WatchOptions) -> Result<WatchHandle> {
        let query = format!(
            "path={}&recursive={}&include_entry={}",
            urlencoding(path),
            opts.recursive,
            opts.include_entry
        );
        let socket = ProcessSocket::connect(
            &self.data_plane.base_url,
            &self.data_plane.token,
            &format!("/runtime/v1/files/watch?{query}"),
        )
        .await?;
        Ok(WatchHandle { socket })
    }
}

fn entry_info(value: &Value) -> EntryInfo {
    EntryInfo {
        name: value
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        path: value
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        file_type: value.get("type").and_then(Value::as_str).map(file_type),
        size: value
            .get("bytes")
            .or_else(|| value.get("size"))
            .and_then(Value::as_u64),
        metadata: value
            .get("metadata")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default(),
    }
}

fn file_type(value: &str) -> FileType {
    match value {
        "file" => FileType::File,
        "dir" | "directory" => FileType::Dir,
        "symlink" => FileType::Symlink,
        other => FileType::Other(other.to_string()),
    }
}

fn filesystem_event(value: Value) -> FilesystemEvent {
    let path = value
        .get("path")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string();
    let event_type = match value
        .get("type")
        .and_then(Value::as_str)
        .unwrap_or("modify")
    {
        "delete" => "remove",
        "modify" => "write",
        other => other,
    }
    .to_string();
    let entry = value.get("file").map(entry_info);
    FilesystemEvent {
        name: path
            .trim_end_matches('/')
            .rsplit('/')
            .next()
            .unwrap_or_default()
            .to_string(),
        path,
        event_type,
        entry,
        raw: value,
    }
}

fn apply_diff_report(value: Value) -> ApplyDiffReport {
    ApplyDiffReport {
        status: value
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        parsed_diff_blocks: usize_value(value.get("parsed_diff_blocks")),
        patches: usize_value(value.get("patches")),
        files: value
            .get("files")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .map(apply_diff_file_summary)
            .collect(),
        summary: apply_diff_summary(value.get("summary").unwrap_or(&Value::Null)),
        applied: string_vec(value.get("applied")),
        failed: value
            .get("failed")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .map(apply_diff_failure)
            .collect(),
        touched: string_vec(value.get("touched")),
        raw: value,
    }
}

fn apply_diff_summary(value: &Value) -> ApplyDiffSummary {
    ApplyDiffSummary {
        requested: usize_value(value.get("requested")),
        applied: usize_value(value.get("applied")),
        failed: usize_value(value.get("failed")),
    }
}

fn apply_diff_file_summary(value: Value) -> ApplyDiffFileSummary {
    ApplyDiffFileSummary {
        path: value
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        source_path: value
            .get("source_path")
            .and_then(Value::as_str)
            .map(str::to_string),
        kind: value
            .get("kind")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        added: usize_value(value.get("added")),
        removed: usize_value(value.get("removed")),
    }
}

fn apply_diff_failure(value: Value) -> ApplyDiffFailure {
    let failed_hunk = value.get("failed_hunk").and_then(|hunk| {
        hunk.as_object().map(|_| ApplyDiffFailedHunk {
            index: usize_value(hunk.get("index")),
            old_start: usize_value(hunk.get("old_start")),
        })
    });
    ApplyDiffFailure {
        path: value
            .get("path")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        error: value
            .get("error")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        failed_hunk,
    }
}

fn usize_value(value: Option<&Value>) -> usize {
    value.and_then(Value::as_u64).unwrap_or(0) as usize
}

fn string_vec(value: Option<&Value>) -> Vec<String> {
    value
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .filter_map(|value| value.as_str().map(str::to_string))
        .collect()
}

fn urlencoding(value: &str) -> String {
    url::form_urlencoded::byte_serialize(value.as_bytes()).collect()
}

fn write_files_payload(files: &[WriteEntry]) -> Value {
    serde_json::json!({
        "files": files
            .iter()
            .map(|file| {
                serde_json::json!({
                    "path": file.path.as_str(),
                    "data_base64": base64::engine::general_purpose::STANDARD.encode(&file.data),
                })
            })
            .collect::<Vec<_>>()
    })
}

fn apply_diff_payload(diff: &str, opts: &ApplyDiffOptions) -> Value {
    let mut payload = serde_json::json!({ "diff": diff });
    if let Some(cwd) = opts.cwd.as_deref().filter(|cwd| !cwd.is_empty()) {
        payload["cwd"] = Value::String(cwd.to_string());
    }
    payload
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{apply_diff_payload, write_files_payload, ApplyDiffOptions, WriteEntry};

    #[test]
    fn write_files_payload_uses_snake_case_base64_entries() {
        assert_eq!(
            write_files_payload(&[
                WriteEntry::new("/tmp/a.txt", "abc"),
                WriteEntry::new("/tmp/b.bin", [0, 1, 2]),
            ]),
            json!({
                "files": [
                    {"path": "/tmp/a.txt", "data_base64": "YWJj"},
                    {"path": "/tmp/b.bin", "data_base64": "AAEC"}
                ]
            })
        );
    }

    #[test]
    fn apply_diff_payload_uses_snake_case_route_body() {
        assert_eq!(
            apply_diff_payload(
                "diff --git a/a.txt b/a.txt",
                &ApplyDiffOptions {
                    cwd: Some("/workspace/app".to_string()),
                }
            ),
            json!({
                "diff": "diff --git a/a.txt b/a.txt",
                "cwd": "/workspace/app"
            })
        );
    }
}
